File: src/main/java/sample/controller/dialogs/RandomSoupDialog.java
package sample.controller.dialogs;

import javafx.beans.property.SimpleStringProperty;
import javafx.scene.control.*;
import javafx.scene.layout.VBox;
import javafx.stage.Window;
import sample.model.SymmetryGenerator;

import java.util.ArrayList;
import java.util.List;

public class RandomSoupDialog extends Dialog {
    private final Slider densitySlider;
    private final ComboBox<String> symmetryCombobox;
    private final ArrayList<CheckBox> statesCheckBox;

    public RandomSoupDialog(int numStates, int density, String symmetry, List<Integer> states) {
        super();

        super.setTitle("Random Soup Settings");
        super.setResizable(true);

        VBox vBox = new VBox();
        vBox.setSpacing(5);

        Label densityLabel = new Label("Density:");
        vBox.getChildren().add(densityLabel);

        // Slider for density
        densitySlider = new Slider();
        densitySlider.setMin(0);
        densitySlider.setMax(100);
        densitySlider.setValue(density);
        densitySlider.valueProperty().addListener((obs, oldval, newVal) ->
                densitySlider.setValue(newVal.intValue()));  // Round to integer
        vBox.getChildren().add(densitySlider);

        densityLabel.textProperty().bind(new SimpleStringProperty("Density: ").concat(
                densitySlider.valueProperty()));

        vBox.getChildren().add(new Label("Symmetry:"));

        // Combobox for symmetries
        symmetryCombobox = new ComboBox<>();
        symmetryCombobox.setValue(symmetry);
        symmetryCombobox.getItems().addAll(SymmetryGenerator.symmetries);
        vBox.getChildren().add(symmetryCombobox);

        // Selector for states
        statesCheckBox = new ArrayList<>();
        for (int i = 1; i < numStates; i++) {
            CheckBox checkBox = new CheckBox(i + "");
            if (states.contains(i)) checkBox.setSelected(true);

            statesCheckBox.add(checkBox);
            vBox.getChildren().add(checkBox);
        }

        // Okay Button
        Button button = new Button("Confirm Settings");
        button.setOnAction(event -> {
            // Close the dialog
            super.setResult(Boolean.TRUE);
            super.close();
        });
        vBox.getChildren().add(button);

        super.getDialogPane().setContent(vBox);

        // Allows closing with close button
        Window window = super.getDialogPane().getScene().getWindow();
        window.setOnCloseRequest(event -> window.hide());
    }

    public int getDensity() {
        return (int) densitySlider.getValue();
    }

    public String getSymmetry() {
        return symmetryCombobox.getValue();
    }

    public ArrayList<Integer> getStates() {
        ArrayList<Integer> states = new ArrayList<>();
        for (int i = 0; i < statesCheckBox.size(); i++) {
            if (statesCheckBox.get(i).isSelected()) states.add(i + 1);
        }

        return states;
    }
}

File: src/main/kotlin/application/controller/dialogs/RandomSoupDialog.kt
package application.controller.dialogs

import application.model.SymmetryGenerator
import javafx.beans.property.SimpleStringProperty
import javafx.beans.value.ObservableValue
import javafx.event.ActionEvent
import javafx.event.EventHandler
import javafx.scene.control.*
import javafx.scene.layout.VBox
import javafx.stage.WindowEvent
import java.lang.Boolean
import kotlin.Any
import kotlin.Int
import kotlin.Number
import kotlin.String

class RandomSoupDialog(numStates: Int, density: Int, symmetry: String, states: List<Int?>) : Dialog<Any?>() {
    private val densitySlider: Slider
    private val symmetryCombobox: ComboBox<String>
    private val statesCheckBox: ArrayList<CheckBox>

    val density: Int
        get() = densitySlider.value.toInt()

    val symmetry: String
        get() = symmetryCombobox.value

    val states: ArrayList<Int>
        get() {
            val states = ArrayList<Int>()
            for (i in statesCheckBox.indices) {
                if (statesCheckBox[i].isSelected) states.add(i + 1)
            }
            return states
        }

    init {
        super.setTitle("Random Soup Settings")
        super.setResizable(true)

        val vBox = VBox()
        vBox.spacing = 5.0

        val densityLabel = Label("Density:")
        vBox.children.add(densityLabel)

        // Slider for density
        densitySlider = Slider()
        densitySlider.min = 0.0
        densitySlider.max = 100.0
        densitySlider.value = density.toDouble()
        densitySlider.valueProperty()
            .addListener { _: ObservableValue<out Number>, _: Number, newVal: Number ->
                densitySlider.value = newVal.toInt().toDouble()
            } // Round to integer
        vBox.children.add(densitySlider)
        densityLabel.textProperty().bind(
            SimpleStringProperty("Density: ").concat(
                densitySlider.valueProperty()
            )
        )
        vBox.children.add(Label("Symmetry:"))

        // Combobox for symmetries
        symmetryCombobox = ComboBox()
        symmetryCombobox.value = symmetry
        symmetryCombobox.items.addAll(*SymmetryGenerator.symmetries)
        vBox.children.add(symmetryCombobox)

        // Selector for states
        statesCheckBox = ArrayList()
        for (i in 1 until numStates) {
            val checkBox = CheckBox(i.toString() + "")
            if (states.contains(i)) checkBox.isSelected = true
            statesCheckBox.add(checkBox)
            vBox.children.add(checkBox)
        }

        // Okay Button
        val button = Button("Confirm Settings")
        button.onAction = EventHandler {
            // Close the dialog
            super.setResult(Boolean.TRUE)
            super.close()
        }
        vBox.children.add(button)
        super.getDialogPane().content = vBox

        // Allows closing with close button
        val window = super.getDialogPane().scene.window
        window.onCloseRequest = EventHandler { window.hide() }
    }
}